go/libraries/doltcore/doltdb/global_table_editor.go

```diff
@@ -0,0 +1,278 @@
+// Copyright 2020 Liquidata, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package doltdb
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"github.com/liquidata-inc/dolt/go/libraries/doltcore/schema"
+)
+
+// GetRootFunc returns the root value that the GlobalTableEditor should be using.
+type GetRootFunc func(ctx context.Context) (*RootValue, error)
+
+// GlobalTableEditor handles all edit operations on a table that may also update other tables. Serves as coordination
+// for LocalTableEditors.
+type GlobalTableEditorImpl struct {
+	DisableForeignKeyChecks bool // If disabled, then all foreign key checks and updates are ignored
+
+	getRoot    GetRootFunc
+	tables     map[string]*LocalTableEditor
+	writeMutex *sync.RWMutex // This mutex is specifically for changes that affect the GlobalTE or all LocalTEs
+}
+
+var globalTableEditor = &GlobalTableEditorImpl{}
+
+func init() {
+	globalTableEditor.tables = make(map[string]*LocalTableEditor)
+	globalTableEditor.writeMutex = &sync.RWMutex{}
+}
+
+// GlobalTableEditor returns the global table editor.
+func GlobalTableEditor() *GlobalTableEditorImpl {
+	return globalTableEditor
+}
+
+// GetTableEditor returns a LocalTableEditor for the given table. Fetches a root from the function provided to SetRoot
+// to get the table and any potential foreign keys.
+func (gte *GlobalTableEditorImpl) GetTableEditor(ctx context.Context, tableName string) (*LocalTableEditor, error) {
+	gte.writeMutex.Lock()
+	defer gte.writeMutex.Unlock()
+
+	return gte.getTableEditorOverride(ctx, tableName, nil, nil)
+}
+
+// GetTableEditor returns a LocalTableEditor using the given table. Fetches a root from the function provided to SetRoot
+// for any potential foreign keys. Any existing LocalTableEditors that reference this table will be updated with the
+// given table, losing any unsaved edits. If a nil table and schema are given, then this functions the same as
+// GetTableEditor.
+func (gte *GlobalTableEditorImpl) GetTableEditorOverride(ctx context.Context, tableName string, t *Table, tableSch schema.Schema) (*LocalTableEditor, error) {
```

Comment:
The concept of an "override" doesn't really make sense here. Or maybe I'm missing something fundamental that you could explain to me. But this doesn't seem like an "override", it's just creating a table editor for the table named, and sometimes throwing away an existing one. From the single use of this method, it's hard for me to understand your intent.

The job of this object should be to manage all open table editors, and create / update / validate them as necessary for clients to use them. If it's necessary to flush / cancel / update an existing table editor for a particular table, then make that explicit with a method of that name. But make that behavior explicit -- or else roll it into GetTableEditor itself, and intelligently handle invalidation there.